lib/records.ncl

```diff
@@ -296,4 +296,57 @@ let error_lib = import "./error.ncl" in
                 }
             )
             'Ok,
+
+  contract = {
+    maxProperties
+      | Number -> Dyn
+      | doc m%"
+        A contract asserting that the value is a record with at most a specific number of properties.
+      "%
+      = fun n =>
+        std.contract.from_validator (fun x =>
+          if !std.is_record x then
+            'Error { message = "expected a record" }
+          else if std.record.length x > n then
+            'Error { message = "record contains more than %{std.string.from_number n} fields" }
+          else
+            'Ok
+        ),
+
+    minProperties
+      | Number -> Dyn
+      | doc m%"
+        A contract asserting that the value is a record with at least a specific number of properties.
+      "%
+      = fun n =>
+        std.contract.from_validator (fun x =>
+          if !std.is_record x then
+            'Error { message = "expected a record" }
+          else if std.record.length x < n then
+            'Error { message = "record contains less than %{std.string.from_number n} fields" }
+          else
+            'Ok
+        ),
+
+    # TODO: replace with FieldsMatch once that's in a nickel release
+    patternFields | String -> Dyn = fun regex =>
+      let is_match = std.string.is_match regex in
+      std.contract.from_validator (fun x =>
+        if !std.is_record x then
+          'Error { message = "expected a record" }
+        else
+          let bad_field_names =
+            std.record.fields x
+            |> std.array.filter (fun k => !(is_match k))
+            |> std.array.map (fun k => m%""${k}""%)
```

Comment:
Nitpicks: we use backticks already quite a bit in Nickel error messages for quoting, and even if we want to use `"`, I'm not sure the `m%"` id paying for itself here.

```suggestion
            |> std.array.map (fun k => "`${k}`")
```